crates/uv-workspace/src/dependency_groups.rs

```diff
@@ -0,0 +1,150 @@
+            name: &'data GroupName,
+            parents: &mut Vec<&'data GroupName>,
+        ) -> Result<(), DependencyGroupError> {
+            let Some(specifiers) = groups.get(name) else {
+                // Missing group
+                let parent_name = parents
+                    .iter()
+                    .last()
+                    .copied()
+                    .expect("parent when group is missing");
+                return Err(DependencyGroupError::GroupNotFound(
+                    name.clone(),
+                    parent_name.clone(),
+                ));
+            };
+
+            // "Dependency Group Includes MUST NOT include cycles, and tools SHOULD report an error if they detect a cycle."
+            if parents.contains(&name) {
+                return Err(DependencyGroupError::DependencyGroupCycle(Cycle(
+                    parents.iter().copied().cloned().collect(),
+                )));
+            }
+
+            // If we already resolved this group, short-circuit.
+            if resolved.contains_key(name) {
+                return Ok(());
+            }
+
+            parents.push(name);
+            let mut requirements = Vec::with_capacity(specifiers.len());
+            for specifier in *specifiers {
+                match specifier {
+                    DependencyGroupSpecifier::Requirement(requirement) => {
+                        match uv_pep508::Requirement::<VerbatimParsedUrl>::from_str(requirement) {
+                            Ok(requirement) => requirements.push(requirement),
+                            Err(err) => {
+                                return Err(DependencyGroupError::GroupParseError(
+                                    name.clone(),
+                                    requirement.clone(),
+                                    Box::new(err),
+                                ));
+                            }
+                        }
+                    }
+                    DependencyGroupSpecifier::IncludeGroup { include_group } => {
+                        resolve_group(resolved, groups, include_group, parents)?;
+                        requirements
+                            .extend(resolved.get(include_group).into_iter().flatten().cloned());
+                    }
+                    DependencyGroupSpecifier::Object(map) => {
+                        warn!(
+                            "Ignoring Dependency Object Specifier referenced by `{name}`: {map:?}"
+                        );
+                    }
+                }
+            }
+            parents.pop();
+
+            resolved.insert(name.clone(), requirements);
+            Ok(())
+        }
+
+        let mut resolved = BTreeMap::new();
+        for name in groups.keys() {
+            let mut parents = Vec::new();
+            resolve_group(&mut resolved, groups, name, &mut parents)?;
+        }
+        Ok(Self(resolved))
```

Comment:
This whole piece moved over from `requires_dist.rs`; no changes. I just wrapped it in a struct so that I could abstract it away and reuse it elsewhere.